include/chainbase/undo_index.hpp

```diff
@@ -0,0 +1,845 @@
+      using pointer = value_type*;
+      using const_pointer = const value_type*;
+
+      static node_ptr to_node_ptr(value_type &value) {
+         return node_ptr{static_cast<Node*>(boost::intrusive::get_parent_from_member(&value, &value_holder<value_type>::_item))};
+      }
+      static const_node_ptr to_node_ptr(const value_type &value) {
+         return const_node_ptr{static_cast<const Node*>(boost::intrusive::get_parent_from_member(&value, &value_holder<value_type>::_item))};
+      }
+      static pointer to_value_ptr(node_ptr n) { return pointer{&static_cast<Node*>(&*n)->_item}; }
+      static const_pointer to_value_ptr(const_node_ptr n) { return const_pointer{&static_cast<const Node*>(&*n)->_item}; }
+
+      static constexpr boost::intrusive::link_mode_type link_mode = boost::intrusive::normal_link;
+   };
+   template<typename Allocator, typename T>
+   using rebind_alloc_t = typename std::allocator_traits<Allocator>::template rebind_alloc<T>;
+
+   template<typename Index>
+   struct index_tag_impl { using type = void; };
+   template<template<typename...> class Index, typename Tag, typename... T>
+   struct index_tag_impl<Index<boost::multi_index::tag<Tag>, T...>> { using type = Tag; };
+   template<typename Index>
+   using index_tag = typename index_tag_impl<Index>::type;
+
+   template<typename Index>
+   using index_key = typename Index::key_from_value_type;
+   template<typename Index>
+   using index_compare = typename Index::compare_type;
+
+   template<typename Tag, typename... Keys>
+   using index_of_tag = boost::mp11::mp_find<boost::mp11::mp_list<index_tag<Keys>...>, Tag>;
+
+   template<typename K, typename Allocator>
+   using hook = offset_node_base<K>;
+
+   template<typename Node, typename Key>
+   using set_base = boost::intrusive::avltree<
+      typename Node::value_type,
+      boost::intrusive::value_traits<offset_node_value_traits<Node, Key>>,
+      boost::intrusive::key_of_value<get_key<index_key<Key>, typename Node::value_type>>,
+      boost::intrusive::compare<index_compare<Key>>>;
+
+   template<typename Node, typename Key>
+   using list_base = boost::intrusive::slist<
+      typename Node::value_type,
+      boost::intrusive::value_traits<offset_node_value_traits<Node, Key>>>;
+
+   template<typename L, typename It, typename Pred, typename Disposer>
+   void remove_if_after_and_dispose(L& l, It it, It end, Pred&& p, Disposer&& d) {
+      for(;;) {
+         It next = it;
+         ++next;
+         if(next == end) break;
+         if(p(*next)) { l.erase_after_and_dispose(it, d); }
+         else { it = next; }
+      }
+   }
+
+   template<typename T, typename Allocator, typename... Keys>
+   class undo_index;
+  
+   template<typename Node, typename Key>
+   struct set_impl : private set_base<Node, Key> {
+     using base_type = set_base<Node, Key>;
+      // Allow compatible keys to match multi_index
+      template<typename K>
+      auto find(K&& k) {
```

Comment:
Unnecessary to have a non-const find in the public interface. If this is needed by the implementation, then make it private. Otherwise, remove.